### PR TITLE
Atolcd bug fixes : Offline mode

### PR DESCRIPTION
**Pull request by ftavin**
This commits handle offline mode for contacts and LayoutGroupIteration.
The first commit also adds tomcat in docker-compose file to help dealing with offline mode dev using docker (instructions in commit message)

### CODE REVIEW

**Comment by osarrat** [comment]
Wow !
That's a great work !

While I will assess that it does functionally the job, could you improve a bit code quality as specified by this Codacy report : https://www.codacy.com/app/osarrat/sigmah/pullRequest?prid=733364 ? (mostly unused imports to remove)

File: docker-compose.yml [hunk 1]
```diff
@@ -13,3 +13,11 @@ mailhog:
...
     - "1025:1025"
     - "8025:8025"
+tomcat:
+  build: ./sigmah-docker/tomcat
+  ports:
+    - "9999:8080"
+  links:
+    - postgresql
+  volumes:
+    - ./target/sigmah-2.2-SNAPSHOT/:/usr/local/tomcat/webapps/sigmah/
```
Comment (ghost) [inline docker-compose.yml]: Can you find a way without having the version `2.2-SNAPSHOT`?

Comment (ghost) [inline docker-compose.yml]: The documentation you put in the commit message is interesting, could you put it in a comment in this file?

File: src/main/java/org/sigmah/shared/dto/element/FlexibleElementDTO.java [hunk 2]
```diff
@@ -297,6 +299,11 @@ public void componentSelected(ButtonEvent ce) {
 						loadAndShowHistory(wrapper.getHistoryButton());
 					}
 				});
+				
+				if (ElementTypeEnum.CONTACT_LIST == this.getElementType() && Profiler.INSTANCE.getApplicationStateManager().getLastState() == ApplicationState.OFFLINE) {
```
Comment (ghost) [inline src/main/java/org/sigmah/shared/dto/element/FlexibleElementDTO.java]: This piece of code

```java
Profiler.INSTANCE.getApplicationStateManager().getLastState() == ApplicationState.OFFLINE
```

is copy-pasted everywhere (sometimes in a method, sometimes not). Can you factorize in a single method somewhere in `Profiler.INSTANCE` or in `ApplicationStateManager`? I guess what you are actually looking for is the flag `ApplicationStateManager#online`

File: src/main/java/org/sigmah/client/ui/presenter/contact/ContactDetailsPresenter.java [hunk 3]
```diff
@@ -578,78 +579,147 @@ public void onRequiredValueChange(RequiredValueEvent event) {
...
-            N10N.infoNotif(I18N.CONSTANTS.infoConfirmation(), I18N.CONSTANTS.saveConfirm());
-            eventBus.fireEvent(new UpdateEvent(UpdateEvent.CONTACT_UPDATE, contactDTO));
-
-            // Checks if there is any update needed to the local project instance.
-            for (final ValueEvent event : valueChanges) {
-              if (event.getSource() instanceof DefaultContactFlexibleElementDTO) {
-                updateCurrentContact(contactDTO, (DefaultContactFlexibleElementDTO) event.getSource(), event.getSingleValue());
-              }
-            }
+		if (!iterationChanges.isEmpty()) {
```
Comment (ghost) [inline src/main/java/org/sigmah/client/ui/presenter/contact/ContactDetailsPresenter.java]: Too long, please break in smaller methods (or maybe extract anonymous class to inner?)

**Review by ghost — CHANGES_REQUESTED** [review]
I had a quick review, also, I second the "really long methods" spotted by Codacy.

**Comment by osarrat** [comment]
Thanks @ftavin for those last modifications. Would it be possible for you to also take into account @numero-six requested changes?

File: docker-tomcat.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+exec docker run -it --rm \
+            --link $(docker-compose ps -q postgresql):postgresql \
+            -v $PWD/target/sigmah-2.2-SNAPSHOT/:/usr/local/tomcat/webapps/sigmah/:ro \
```
Comment (ghost) [inline docker-tomcat.sh]: Please, remove hard-coded `2.2-SNAPSHOT`.

You can start your script with something like

```bash
# gives SIGMAH_VERSION="2.2-SNAPSHOT"
SIGMAH_VERSION=`mvn --quiet -Dexec.executable='echo' -Dexec.args='${project.version}' --non-recursive exec:exec`
```

See also https://stackoverflow.com/questions/3545292/how-to-get-maven-project-version-to-the-bash-command-line

**Comment by osarrat** [comment]
Hi @numero-six 
Is this PR mergeable according to you ?
Best

**Comment by ghost** [comment]
Codacy has good points but nothing like an error.

Does it ends the 2.2 developments? If yes I would be happy to tag the final `2.2`.

I'm asking because 2.3 development has started but the roadmap show that there are remaining issues.

**Comment by osarrat** [comment]
Thanks @numero-six ! I merge it now.
Regarding making it the final v2.2, I'll come back on this issue next week.

**Comment by osarrat** [comment]
Hi @numero-six and @ftavin 

I wish to discuss with @ftavin on whether this v2.2-rc6 should be the final v2.2 or not.
So the point is not concluded today, and since we need to publish this v2.2-rc6 for a usage by CARE today, we can consider that a v2.2-rc6 will be released.
We will see later on if we releease this v2.2-rc6 also as the final v2.2.

Best !

**Comment by ghost** [comment]
Hi! `2.2-rc6` tag published

**Comment by osarrat** [comment]
Great, thank you !

**Pull request merged**